README.md
ZF Content Validation
=====================

[![Build Status](https://secure.travis-ci.org/zfcampus/zf-content-validation.svg?branch=master)](https://secure.travis-ci.org/zfcampus/zf-content-validation)
[![Coverage Status](https://coveralls.io/repos/github/zfcampus/zf-content-validation/badge.svg?branch=master)](https://coveralls.io/github/zfcampus/zf-content-validation?branch=master)

Introduction
------------

Zend Framework module for automating validation of incoming input.

Allows the following:

- Defining named input filters.
- Mapping named input filters to named controller services.
- Returning an `ApiProblemResponse` with validation error messages on invalid input.

Requirements
------------
  
Please see the [composer.json](composer.json) file.

Installation
------------

Run the following `composer` command:

```console
$ composer require zfcampus/zf-content-validation
```

Alternately, manually add the following to your `composer.json`, in the `require` section:

```javascript
"require": {
    "zfcampus/zf-content-validation": "^1.4"
}
```

And then run `composer update` to ensure the module is installed.

Finally, add the module name to your project's `config/application.config.php` under the `modules`
key:

```php
return [
    /* ... */
    'modules' => [
        /* ... */
        'ZF\ContentValidation',
    ],
    /* ... */
];
```

Configuration
-------------

### User Configuration

This module utilizes two user level configuration keys `zf-content-validation` and also
`input_filter_specs` (named such that this functionality can be moved into ZF2 in the future).

#### Service Name key

The `zf-content-validation` key is a mapping between controller service names as the key, and the
value being an array of mappings that determine which HTTP method to respond to and what input
filter to map to for the given request.  The keys for the mapping can either be an HTTP method that
accepts a request body (i.e., `POST`, `PUT`, `PATCH`, or `DELETE`), or it can be the word
`input_filter`. The value assigned for the `input_filter` key will be used in the case that no input
filter is configured for the current HTTP request method.

Example where there is a default as well as a POST filter:

```php
'zf-content-validation' => [
    'Application\Controller\HelloWorld' => [
        'input_filter' => 'Application\Controller\HelloWorld\Validator',
        'POST' => 'Application\Controller\HelloWorld\CreationValidator',
    ],
],
```

In the above example, the `Application\Controller\HelloWorld\Validator` service will be selected for
`PATCH`, `PUT`, or `DELETE` requests, while the `Application\Controller\HelloWorld\CreationValidator`will be selected for `POST` requests.

Starting in version 1.1.0, two additional keys can be defined to affect application validation
behavior:

- `use_raw_data`: if NOT present, raw data is ALWAYS injected into the "BodyParams" container (defined
  by zf-content-negotiation).  If this key is present and a boolean false, then the validated,
  filtered data from the input filter will be used instead.

- `allows_only_fields_in_filter`: if present, and `use_raw_data` is boolean false, the value of this
  flag will define whether or not additional fields present in the payload will be merged with the
  filtered data.
  
- `remove_empty_data`: Should we remove empty data from received data?
  - If no `remove_empty_data` flag is present, do nothing - use data as is
  - If `remove_empty_data` flag is present AND is boolean true, then remove
    empty data from current data array
  - Does not remove empty data if keys matched received data

> ### Validating GET requests
>
> - **Since 1.3.0.**
>
> Starting in 1.3.0, you may also specify `GET` as an HTTP method, mapping it to
> an input filter in order to validate your query parameters. Configuration is
> exactly as described in the above section.
>
> This feature is only available when manually configuring your API; it is not
> exposed in the Admin UI.

> ### Validating collection requests
>
> - **Since 1.5.0**
>
> Starting in 1.5.0, you may specify any of:
>
> - `POST_COLLECTION`
> - `PUT_COLLECTION`
> - `PATCH_COLLECTION`
>
> as keys. These will then be used specifically with the given HTTP method, but
> only on requests matching the collection endpoint.

> ### Validating DELETE requests
>
> - **Since 1.6.0**
>
> Starting in 1.6.0, you may specify each of the following keys for input
> filters:
>
> - `DELETE`
> - `DELETE_COLLECTION`
>
> The input filter associated with the key will be used to validate data sent in
> the request body.

#### input_filter_spec

`input_filter_spec` is for configuration-driven creation of input filters.  The keys for this array
will be a unique name, but more often based off the service name it is mapped to under the
`zf-content-validation` key.  The values will be an input filter configuration array, as is
described in the ZF2 manual [section on input
filters](http://zf2.readthedocs.org/en/latest/modules/zend.input-filter.intro.html).

Example:

```php
'input_filter_specs' => [
    'Application\Controller\HelloWorldGet' => [
        0 => [
            'name' => 'name',
            'required' => true,
            'filters' => [
                0 => [
                    'name' => 'Zend\Filter\StringTrim',
                    'options' => [],
                ],
            ],
            'validators' => [],
            'description' => 'Hello to name',
            'allow_empty' => false,
            'continue_if_empty' => false,
        ],
    ],
```

### System Configuration

The following configuration is defined by the module in order to function within a ZF2 application.

```php
namespace ZF\ContentValidation;

use Zend\InputFiler\InputFilterAbstractServiceFactory;
use Zend\ServiceManager\Factory\InvokableFactory;

return [
    'controller_plugins' => [
        'aliases' => [
            'getinputfilter' => InputFilter\InputFilterPlugin::class,
            'getInputfilter' => InputFilter\InputFilterPlugin::class,
            'getInputFilter' => InputFilter\InputFilterPlugin::class,
        ],
        'factories' => [
            InputFilter\InputFilterPlugin::class => InvokableFactory::class,
        ],
    ],
    'input_filters' => [
        'abstract_factories' => [
            InputFilterAbstractServiceFactory::class,
        ],
    ],
    'service_manager' => [
        'factories' => [
            ContentValidationListener::class => ContentValidationListenerFactory::class,
        ],
    ],
    'validators' => [
        'factories' => [
            'ZF\ContentValidation\Validator\DbRecordExists' => Validator\Db\RecordExistsFactory::class,
            'ZF\ContentValidation\Validator\DbNoRecordExists' => Validator\Db\NoRecordExistsFactory::class,
        ],
    ],
];
```

ZF Events
---------

### Listeners

#### ZF\ContentValidation\ContentValidationListener

This listener is attached to the `MvcEvent::EVENT_ROUTE` event at priority `-650`.  Its purpose is
to utilize the `zf-content-validation` configuration in order to determine if the current request's
selected controller service name has a configured input filter.  If it does, it will traverse the
mappings from the configuration file to create the appropriate input filter (from configuration or
the Zend Framework 2 input filter plugin manager) in order to validate the incoming data.  This
particular listener utilizes the data from the `zf-content-negotiation` data container in order to
get the deserialized content body parameters.

### Events

#### ZF\ContentValidation\ContentValidationListener::EVENT_BEFORE_VALIDATE

This event is emitted by `ZF\ContentValidation\ContentValidationListener::onRoute()`
(described above) in between aggregating data to validate and determining the
input filter, and the actual validation of data. Its purpose is to allow users:

- the ability to manipulate input filters. 
- to modify the data set to validate (available since 1.4.0).

As an example, you might want to validate an identifier provided via the URI,
and matched during routing. You may do this as follows:

```php
$events->listen(ContentValidationListener::EVENT_BEFORE_VALIDATE, function ($e) {
    if ($e->getController() !== MyRestController::class) {
        return;
    }

    $matches = $e->getRouteMatch();
    $data = $e->getParam('ZF\ContentValidation\ParameterData') ?: [];
    $data['id'] = $matches->getParam('id');
    $e->setParam('ZF\ContentValidation\ParameterData', $data);
});
```

ZF Services
-----------

### Controller Plugins

#### ZF\ContentValidation\InputFilter\InputFilterPlugin (aka getInputFilter)

This plugin is available to Zend Framework 2 controllers. When invoked (`$this->getInputFilter()` or
`$this->plugin('getinputfilter')->__invoke()`), it returns whatever is in the MVC event parameter
`ZF\ContentValidation\InputFilter`, returning null for any value that is not an implementation of
`Zend\InputFilter\InputFilter`.

### Service

#### Zend\InputFilter\InputFilterAbstractServiceFactory

This abstract factory is responsible for creating and returning an appropriate input filter given
a name and the configuration from the top-level key `input_filter_specs`. It is registered with
`Zend\InputFilter\InputFilterPluginManager`.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file, in reverse chronological order by release.

## 1.9.0 - TBD

### Added

- Nothing.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 1.8.1 - TBD

### Added

- Nothing.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 1.8.0 - 2019-03-05

### Added

- [#106](https://github.com/zfcampus/zf-content-validation/pull/106) adds support for PHP 7.3.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- [#106](https://github.com/zfcampus/zf-content-validation/pull/106) removes support for zend-stdlib v2 releases.

### Fixed

- Nothing.

## 1.7.1 - 2019-03-05

### Added

- Nothing.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- [#105](https://github.com/zfcampus/zf-content-validation/pull/105) fixes default behaviour when the configuration flag `remove_empty_data`
  is not set; in that scenario, data will not change as described in the
  documentation. This fixes a BC break introduced in version 1.7.0.

## 1.7.0 - 2019-02-26

### Added

- [#102](https://github.com/zfcampus/zf-content-validation/pull/102) adds the configuration flag `remove_empty_data`. If set and boolean 
  `true`, it causes removal of empty key/value pairs from filtered input data.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 1.6.1 - 2018-08-13

### Added

- Nothing.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- [#99](https://github.com/zfcampus/zf-content-validation/pull/99) fixes how the `ContentValidationListener` chooses an input filter
  on `DELETE` requests. Starting in 1.6.0, it erroneously would fall back to the
  default input filter if no `DELETE` or `DELETE_COLLECTION` input filter was
  found. However, this was a backwards incompatible change from previous
  versions, which would not apply the default input filter for `DELETE`
  requests. As such, a new check was added to provide the previous behavior in
  such situations.

## 1.6.0 - 2018-08-02

### Added

- [#98](https://github.com/zfcampus/zf-content-validation/pull/98) adds support for specifying controller-specific input filters for
  `DELETE` requests, using the keys `DELETE` and `DELETE_COLLECTION`. The input
  filter associated with the key will be applied to the request body.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 1.5.0 - 2018-05-29

### Added

- [#96](https://github.com/zfcampus/zf-content-validation/pull/96) adds the ability to specify different input filters for _collections_ (versus
  entities). These may be indicated using `<METHOD>_COLLECTION` within your resource input filter configuration.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 1.4.0 - 2018-05-07

### Added

- [#95](https://github.com/zfcampus/zf-content-validation/pull/95) adds support for PHP 7.1 and 7.2.

### Changed

- [#93](https://github.com/zfcampus/zf-content-validation/pull/93) modifies `ContentValidationListener::onRoute()` to set the event parameter
  `ZF\ContentValidation\ParameterData` with the data it has aggregated prior to emitting the event
  `ContentValidationListener::EVENT_BEFORE_VALIDATE`. It also now pulls that parameter immediately
  prior to validating data, merging its value with the original data aggregated, and using the
  result when validating. This feature allows you to manipulate the data being validated prior
  to validation; one use case is to inject route matches into the data set to validate.

### Deprecated

- Nothing.

### Removed

- [#95](https://github.com/zfcampus/zf-content-validation/pull/95) removes support for HHVM.

### Fixed

- Nothing.

## 1.3.6 - 2017-11-06

### Added

- Nothing.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- [#88](https://github.com/zfcampus/zf-content-validation/issues/88) fixes an
  issue when handling PATCH requests in JSON formats that include a blank field
  name. Prior to the patch, these would raise a 400 error with a stack trace;
  with the fix, we now issue a 400 error with the message 'Unrecognized field
  ""'.

## 1.3.5 - 2016-08-18

### Added

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- [#83](https://github.com/zfcampus/zf-content-validation/pull/83) fixes how
  input filters are processed when performing GET collections to REST endpoints.
  This ability was introduce in 1.3.0, but failed to take into account that
  query string arguments cannot properly represent multiple entities, which led
  to any input filter for a `ZF\Rest\Resource::fetchAll()` operation to be
  invalid. The change introduced treats such input filters such that they
  do not operate as a `CollectionInputFilter`; as such, they should only be used
  to retrieve values that will be used to sort or filter the collection
  returned.
- [#84](https://github.com/zfcampus/zf-content-validation/pull/84) fixes how
  unknown fields are pulled from `CollectionInputFilter` instances when
  reporting their presence for Problem Details. Previously, it would report only
  the index of the collection set, and not the values from each; with this
  release, it now reports the which sets had unknown values, along with the
  unexpected values detected.

## 1.3.4 - 2016-08-04

### Added

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- [#81](https://github.com/zfcampus/zf-content-validation/pull/81) fixes an

src/ContentValidationListener.php
<?php
/**
 * @license   http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
 * @copyright Copyright (c) 2014-2018 Zend Technologies USA Inc. (http://www.zend.com)
 */

namespace ZF\ContentValidation;

use Zend\EventManager\EventManager;
use Zend\EventManager\EventManagerAwareInterface;
use Zend\EventManager\EventManagerInterface;
use Zend\EventManager\ListenerAggregateInterface;
use Zend\EventManager\ListenerAggregateTrait;
use Zend\Http\Request as HttpRequest;
use Zend\Http\Response;
use Zend\InputFilter\CollectionInputFilter;
use Zend\InputFilter\Exception\InvalidArgumentException as InputFilterInvalidArgumentException;
use Zend\InputFilter\InputFilterInterface;
use Zend\InputFilter\UnknownInputsCapableInterface;
use Zend\Mvc\MvcEvent;
use Zend\Mvc\Router\RouteMatch as V2RouteMatch;
use Zend\Router\RouteMatch;
use Zend\ServiceManager\ServiceLocatorInterface;
use Zend\Stdlib\ArrayUtils;
use ZF\ApiProblem\ApiProblem;
use ZF\ApiProblem\ApiProblemResponse;
use ZF\ContentNegotiation\ParameterDataContainer;

class ContentValidationListener implements ListenerAggregateInterface, EventManagerAwareInterface
{
    use ListenerAggregateTrait;

    const EVENT_BEFORE_VALIDATE = 'contentvalidation.beforevalidate';

    /**
     * @var array
     */
    protected $config = [];

    protected $events;

    /**
     * @var ServiceLocatorInterface
     */
    protected $inputFilterManager;

    /**
     * Cache of input filter service names/instances
     *
     * @var array
     */
    protected $inputFilters = [];

    /**
     * @var array
     */
    protected $methodsWithoutBodies = [
        'GET',
        'HEAD',
        'OPTIONS',
    ];

    /**
     * Map of REST controllers => route identifier names
     *
     * Used to determine if we have a collection or an entity, for purposes of validation.
     *
     * @var array
     */
    protected $restControllers;

    /**
     * @param array $config
     * @param null|ServiceLocatorInterface $inputFilterManager
     * @param array $restControllers
     */
    public function __construct(
        array $config = [],
        ServiceLocatorInterface $inputFilterManager = null,
        array $restControllers = []
    ) {
        $this->config               = $config;
        $this->inputFilterManager   = $inputFilterManager;
        $this->restControllers      = $restControllers;

        if (isset($config['methods_without_bodies']) && is_array($config['methods_without_bodies'])) {
            foreach ($config['methods_without_bodies'] as $method) {
                $this->addMethodWithoutBody($method);
            }
        }
    }

    /**
     * Set event manager instance
     *
     * Sets the event manager identifiers to the current class, this class, and
     * the resource interface.
     *
     * @param  EventManagerInterface $events
     * @return ContentValidationListener
     */
    public function setEventManager(EventManagerInterface $events)
    {
        $events->addIdentifiers([
            get_class($this),
            __CLASS__,
            self::EVENT_BEFORE_VALIDATE
        ]);
        $this->events = $events;

        return $this;
    }

    /**
     * Retrieve event manager
     *
     * Lazy-instantiates an EM instance if none provided.
     *
     * @return EventManagerInterface
     */
    public function getEventManager()
    {
        if (! $this->events) {
            $this->setEventManager(new EventManager());
        }
        return $this->events;
    }

    /**
     * @see   ListenerAggregateInterface
     * @param EventManagerInterface $events
     * @param int $priority
     */
    public function attach(EventManagerInterface $events, $priority = 1)
    {
        // trigger after authentication/authorization and content negotiation
        $this->listeners[] = $events->attach(MvcEvent::EVENT_ROUTE, [$this, 'onRoute'], -650);
    }

    /**
     * Attempt to validate the incoming request
     *
     * If an input filter is associated with the matched controller service,
     * attempt to validate the incoming request, and inject the event with the
     * input filter, as the "ZF\ContentValidation\InputFilter" parameter.
     *
     * Uses the ContentNegotiation ParameterDataContainer to retrieve parameters
     * to validate, and returns an ApiProblemResponse when validation fails.
     *
     * Also returns an ApiProblemResponse in cases of:
     *
     * - Invalid input filter service name
     * - Missing ParameterDataContainer (i.e., ContentNegotiation is not registered)
     *
     * @param MvcEvent $e
     * @return null|ApiProblemResponse
     */
    public function onRoute(MvcEvent $e)
    {
        $request = $e->getRequest();
        if (! $request instanceof HttpRequest) {
            return;
        }

        $routeMatches = $e->getRouteMatch();
        if (! ($routeMatches instanceof RouteMatch || $routeMatches instanceof V2RouteMatch)) {
            return;
        }

        $controllerService = $routeMatches->getParam('controller', false);
        if (! $controllerService) {
            return;
        }

        $method = $request->getMethod();
        $dataContainer = $e->getParam('ZFContentNegotiationParameterData', false);
        if (! $dataContainer instanceof ParameterDataContainer) {
            return new ApiProblemResponse(
                new ApiProblem(
                    500,
                    'ZF\\ContentNegotiation module is not initialized; cannot validate request'
                )
            );
        }

        $data = $receivedData = in_array($method, $this->methodsWithoutBodies)
            ? $dataContainer->getQueryParams()
            : $dataContainer->getBodyParams();

        if (null === $data || '' === $data) {
            $data = [];
        }

        $isCollection = $this->isCollection($controllerService, $data, $routeMatches, $request);

        $inputFilterService = $this->getInputFilterService($controllerService, $method, $isCollection);

        if (! $inputFilterService) {
            return;
        }

        if (! $this->hasInputFilter($inputFilterService)) {
            return new ApiProblemResponse(
                new ApiProblem(
                    500,
                    sprintf('Listed input filter "%s" does not exist; cannot validate request', $inputFilterService)
                )
            );
        }

        $files = $request->getFiles();
        if (! $isCollection && 0 < count($files)) {
            // File uploads are not validated for collections; impossible to
            // match file fields to discrete sets
            $data = ArrayUtils::merge($data, $files->toArray(), true);
        }

        $inputFilter = $this->getInputFilter($inputFilterService);

        if ($isCollection && ! in_array($method, $this->methodsWithoutBodies)) {
            $collectionInputFilter = new CollectionInputFilter();
            $collectionInputFilter->setInputFilter($inputFilter);
            $inputFilter = $collectionInputFilter;
        }

        $e->setParam('ZF\ContentValidation\InputFilter', $inputFilter);
        $e->setParam('ZF\ContentValidation\ParameterData', $data);

        $currentEventName = $e->getName();
        $e->setName(self::EVENT_BEFORE_VALIDATE);

        $events = $this->getEventManager();
        $results = $events->triggerEventUntil(function ($result) {
            return ($result instanceof ApiProblem
                || $result instanceof ApiProblemResponse
            );
        }, $e);
        $e->setName($currentEventName);

        $last = $results->last();

        if ($last instanceof ApiProblem) {
            $last = new ApiProblemResponse($last);
        }

        if ($last instanceof ApiProblemResponse) {
            return $last;
        }

        $data = ArrayUtils::merge($data, $e->getParam('ZF\ContentValidation\ParameterData'), true);

        $inputFilter->setData($data);

        $status = ($request->isPatch())
            ? $this->validatePatch($inputFilter, $data, $isCollection)
            : $inputFilter->isValid();

        if ($status instanceof ApiProblemResponse) {
            return $status;
        }

        // Invalid? Return a 422 response.
        if (false === $status) {
            return new ApiProblemResponse(
                new ApiProblem(422, 'Failed Validation', null, null, [
                    'validation_messages' => $inputFilter->getMessages(),
                ])
            );
        }

        // Should we use the raw data vs. the filtered data?
        // - If no `use_raw_data` flag is present, always use the raw data, as
        //   that was the default experience starting in 1.0.
        // - If the flag is present AND is boolean true, that is also
        //   an indicator that the raw data should be present.
        $useRawData = $this->useRawData($controllerService);
        if (! $useRawData) {
            $data = $inputFilter->getValues();
        }

        // Should we remove empty data from received data?
        // - If no `remove_empty_data` flag is present, do nothing - use data as is
        // - If `remove_empty_data` flag is present AND is boolean true, then remove
        //   empty data from current data array
        // - Does not remove empty data if keys matched received data
        $removeEmptyData = $this->shouldRemoveEmptyData($controllerService);
        if ($removeEmptyData) {
            $data = $this->removeEmptyData($data, $receivedData);
        }

        // If we don't have an instance of UnknownInputsCapableInterface, or no
        // unknown data is in the input filter, at this point we can just
        // set the current data into the data container.
        if (! $inputFilter instanceof UnknownInputsCapableInterface
            || ! $inputFilter->hasUnknown()
        ) {
            $dataContainer->setBodyParams($data);
            return;
        }

        $unknown    = $inputFilter->getUnknown();

        if ($this->allowsOnlyFieldsInFilter($controllerService)) {
            if ($inputFilter instanceof CollectionInputFilter) {
                $unknownFields = [];
                foreach ($unknown as $key => $fields) {
                    $unknownFields[] = '[' . $key . ': ' . implode(', ', array_keys($fields)) . ']';
                }
                $fields = implode(', ', $unknownFields);
            } else {
                $fields = implode(', ', array_keys($unknown));
            }
            $detail  = sprintf('Unrecognized fields: %s', $fields);
            $problem = new ApiProblem(Response::STATUS_CODE_422, $detail);

            return new ApiProblemResponse($problem);
        }

        // The raw data already contains unknown inputs, so no need to merge
        // them with the data.
        if ($useRawData) {
            $dataContainer->setBodyParams($data);
            return;
        }

        // When not using raw data, we merge the unknown data with the
        // validated data to get the full set of input.
        $dataContainer->setBodyParams(array_merge($data, $unknown));
    }

    /**
     * Add HTTP Method without body content
     *
     * @param string $method
     */
    public function addMethodWithoutBody($method)
    {
        $this->methodsWithoutBodies[] = $method;
    }

    /**
     * @param string $controllerService
     * @return boolean
     */
    protected function allowsOnlyFieldsInFilter($controllerService)
    {
        if (isset($this->config[$controllerService]['allows_only_fields_in_filter'])) {
            return true === $this->config[$controllerService]['allows_only_fields_in_filter'];
        }

        return false;
    }

    /**
     * @param $controllerService
     * @return bool
     */
    protected function useRawData($controllerService)
    {
        if (! isset($this->config[$controllerService]['use_raw_data'])
            || (isset($this->config[$controllerService]['use_raw_data'])
                && $this->config[$controllerService]['use_raw_data'] === true)
        ) {
            return true;
        }
        return false;
    }

    /**
     * @param string $controllerService
     * @return bool
     */
    protected function shouldRemoveEmptyData($controllerService)
    {
        if (isset($this->config[$controllerService]['remove_empty_data'])
            && $this->config[$controllerService]['remove_empty_data'] === true
        ) {
            return true;
        }
        return false;
    }

    /**
     * @param array $data Data to filter null values from
     * @param array $compareTo Original data, send along to preserve
     *     keys/values in $data which are intentional
     * @return array
     */
    protected function removeEmptyData(array $data, array $compareTo = [])
    {
        /**
         * Callback for array_filter() to remove null values (array_filter() removes 'false' values)
         *
         * @param mixed $value
         * @param null  $key
         * @return bool
         */
        $removeNull = function ($value, $key = null) use ($compareTo) {
            // If comparison array is empty, do a straight comparison
            if (empty($compareTo)) {
                return null !== $value;
            }

            // If key exists in comparison array, the 'null' value is on purpose, leave as is
            if (array_key_exists($key, $compareTo)) {
                return true;
            }

            return null !== $value;
        };

        $data = array_filter($data, $removeNull, ARRAY_FILTER_USE_BOTH);

        if (empty($data)) {
            return $data;
        }

        foreach ($data as $key => $value) {
            if (! is_array($value)
                && (! empty($value) || is_bool($value) && ! in_array($key, $compareTo))
            ) {
                continue;
            }

            if (! is_array($value)) {
                unset($data[$key]);
                continue;
            }

            if (empty(array_filter($value, $removeNull, ARRAY_FILTER_USE_BOTH))) {
                unset($data[$key]);
                continue;
            }

            $tmpValue = array_key_exists($key, $compareTo) && is_array($compareTo[$key])
                ? $this->removeEmptyData($value, $compareTo[$key])
                : $this->removeEmptyData($value);

            // Additional check to ensure it's not an empty recursive result
            if (empty(array_filter($tmpValue, $removeNull, ARRAY_FILTER_USE_BOTH))) {
                unset($data[$key]);
                continue;
            }

            $data[$key] = $tmpValue;
        }

        return $data;
    }

    /**
     * Retrieve the input filter service name
     *
     * Test first to see if we have a method-specific input filter, and
     * secondarily for a general one.
     *
     * If neither are present, return boolean false.
     *
     * @param  string $controllerService
     * @param  string $method
     * @param  bool $isCollection
     * @return string|false
     */
    protected function getInputFilterService($controllerService, $method, $isCollection)
    {
        if ($isCollection && isset($this->config[$controllerService][$method . '_COLLECTION'])) {
            return $this->config[$controllerService][$method . '_COLLECTION'];
        }

        if (isset($this->config[$controllerService][$method])) {
            return $this->config[$controllerService][$method];
        }

        if ($method === 'DELETE' || in_array($method, $this->methodsWithoutBodies)) {
            return false;
        }

        if (isset($this->config[$controllerService]['input_filter'])) {
            return $this->config[$controllerService]['input_filter'];
        }

        return false;
    }

    /**
     * Determine if we have an input filter matching the service name
     *
     * @param string $inputFilterService
     * @return bool
     */
    protected function hasInputFilter($inputFilterService)
    {
        if (array_key_exists($inputFilterService, $this->inputFilters)) {
            return true;
        }

        if (! $this->inputFilterManager
            || ! $this->inputFilterManager->has($inputFilterService)
        ) {
            return false;
        }

        $inputFilter = $this->inputFilterManager->get($inputFilterService);
        if (! $inputFilter instanceof InputFilterInterface) {
            return false;
        }

        $this->inputFilters[$inputFilterService] = $inputFilter;
        return true;
    }

    /**
     * Retrieve the named input filter service
     *
     * @param string $inputFilterService
     * @return InputFilterInterface
     */
    protected function getInputFilter($inputFilterService)
    {
        return $this->inputFilters[$inputFilterService];
    }

    /**
     * Does the request represent a collection?
     *
     * @param string $serviceName
     * @param array $data
     * @param RouteMatch|V2RouteMatch $matches
     * @param HttpRequest $request
     * @return bool
     */
    protected function isCollection($serviceName, $data, $matches, HttpRequest $request)
    {
        if (! array_key_exists($serviceName, $this->restControllers)) {
            return false;
        }

        if ($request->isPost() && (empty($data) || ArrayUtils::isHashTable($data))) {
            return false;
        }

        $identifierName = $this->restControllers[$serviceName];
        if ($matches->getParam($identifierName) !== null) {
            return false;
        }

        return (null === $request->getQuery($identifierName, null));
    }

    /**
     * Validate a PATCH request
     *
     * @param InputFilterInterface $inputFilter
     * @param array|object $data
     * @param bool $isCollection
     * @return bool|ApiProblemResponse
     */
    protected function validatePatch(InputFilterInterface $inputFilter, $data, $isCollection)
    {
        if ($isCollection) {
            $validationGroup = $data;
            foreach ($validationGroup as &$subData) {
                $subData = array_keys($subData);
            }
        } else {
            $validationGroup = array_keys($data);
        }

        try {
            $inputFilter->setValidationGroup($validationGroup);
            return $inputFilter->isValid();
        } catch (InputFilterInvalidArgumentException $ex) {
            $pattern = '/expects a list of valid input names; "(?P<field>[^"]*)" was not found/';
            $matched = preg_match($pattern, $ex->getMessage(), $matches);
            if ($matched === 0) {
                return new ApiProblemResponse(
                    new ApiProblem(400, $ex)
                );
            }

            return new ApiProblemResponse(
                new ApiProblem(400, 'Unrecognized field "' . $matches['field'] . '"')
            );
        }
    }
}
